### PR TITLE
Test for the new constant values in Michelson and constant fixups

### PR DESCRIPTION
**Pull request by mariari**

### CODE REVIEW

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 1]
```diff
@@ -444,7 +444,23 @@ onPairGen1 op f =
...
+    ( \instr1 -> do
+        addExpanded instr1
+        case val instr1 of
+          Env.Constant _ -> pure Env.Nop
+          _ -> pure (val instr1)
+    )
+
+--------------------------------------------------------------------------------
+-- On Argument Functions
+--------------------------------------------------------------------------------
```
Comment (mariari) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:459]: Please review this entire section

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 2]
```diff
@@ -58,6 +58,8 @@ inst (Types.Ann _usage ty t) =
   case t of
     Ann.Var symbol -> var symbol
     Ann.AppM fun a -> appM fun a
+    Ann.UnitM -> undefined
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:61]: this should probably not be undefined? or will this never get called? at least `error` if so?

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 3]
```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
...
+
+------------------------------------------------------------
+-- Generator functions
+------------------------------------------------------------
+
+onTwoArgsGen :: OnTerm2Gen m Protect Env.Expanded
+onTwoArgsGen applyOperation typ arguments = do
   -- last argument evaled first
-  v <- traverse (protect . (inst >=> promoteTopStack)) (reverse instrs)
+  v <- traverse (protect . (inst >=> promoteTopStack)) (reverse arguments)
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:471]: we could abstract this, there's code in common between `onTwoArgsGen` and `onOneArgGen`

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 4]
```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
...
+onOneArgsNoConst ::
+  Env.Reduction m => Instr.ExpandedOp -> Types.Type -> [Types.RawTerm] -> m Env.Expanded
+onOneArgsNoConst op =
+  onOneArgGen (noOneConstantCase op)
+
+onTwoArgs :: OnTerm2 m (V.Value' Types.Op) Env.Expanded
+onTwoArgs op f =
+  onTwoArgsGen
+    ( \instr2 instr1 -> do
+        -- May be the wrong order?
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:516]: why does it matter? we're just calling `allConstants`, no?

Comment (cwgoes) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:516]: I don't really see why we need to make a temporary list here at all tbh

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 5]
```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
...
+  -- add the op
+  addInstr op
+  -- return a nop to signal an already added output
+  pure Env.Nop
+
+noOneConstantCase ::
+  Env.Ops m => Instr.ExpandedOp -> Protect -> m Env.Expanded
+noOneConstantCase op instr1 = do
+  addExpanded instr1
+  -- copyAndDrop 1
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:559]: what does this mean?

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 6]
```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
...
+    )
+
+------------------------------------------------------------
+-- Helper predicate/abstractions for the on functions
+------------------------------------------------------------
+
+noTwoConstantCase ::
+  Env.Ops m => Instr.ExpandedOp -> Protect -> Protect -> m Env.Expanded
+noTwoConstantCase op instr2 instr1 = do
+  -- May be the wrong order?
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:546]: this will result in the stack `res1 : res2 : ...` after execution presumably, is that what we want?

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 7]
```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
...
+  let instrs = [instr2, instr1]
+  -- add the protected instructions
+  traverse_ addExpanded instrs
+  -- add the op
+  addInstr op
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:551]: does this appropriately alter the virtual stack?

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 8]
```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
...
+  -- return a nop to signal an already added output
+  pure Env.Nop
+
+noOneConstantCase ::
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:555]: why do we need `noTwoConstantCase` and `noOneConstantCase`, seems like they could be one function?

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 9]
```diff
@@ -745,12 +822,18 @@ valToBool _ = error "called valToBool on a non Michelson Bool"
...
+  -- TODO ∷ determine if this should be a constant, they should all be in
+  -- vstack...
+  -- also once we do this, we can remove the StackOps effect of this function!
   | isNonFunctionPrim prim = do
-    prim <- Env.Expanded <$> primToArgs prim ty
-    consVal prim ty
-    pure prim
+    primToArgs prim ty
+    -- cons to show we have it on the stack
+    consVal Env.Nop ty
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:834]: honestly I just find this pattern confusing (having to use `Nop` like this when we're not actually emitting it as an instr), I guess it's a topic for a later refactor though

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 10]
```diff
@@ -173,9 +177,9 @@ var symb = do
       pure (Env.Curr lamPartial)
     Just (VStack.Position (VStack.Usage usage _saved) index)
       | usage == one ->
-        Env.Expanded <$> moveToFront index
+        Env.Nop <$ moveToFront index
```
Comment (cwgoes) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:180]: I'm confused as to why this change is in this PR, was it just wrong? What does this have to do with the other changes?

**Review by cwgoes — CHANGES_REQUESTED** [review]
The test-case looks correct; confused by some of the code changes.

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 4 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:516]: For abstraction purposes we did this in the past, as we send it into the all check function... basically to scale it up to N arguments with the same pattern... since haskell's type system makes it hard to abstract

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 5 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:559]: Comment was from old code, that I propagated, was something we did that was wrong, but still may serve to be helpful, I can delete if you wish

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 6 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:546]: Yes, that is what we want, at least that is what we've had in the past.

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 8 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:555]: You are correct, abstracting it out, leads to another optimization!

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 10 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:180]: This makes 0 difference, in the past you've mentioned adding `Nop` as you found functions returning values and doing stack operations confusion. This would be another consistent place to add it

Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:180]: Like with many changes, I found this while reasoning through the backend again.

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 9 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:834]: Yeah, I could put an expanded from the `primToArgs`, however since this pushes to the stack, we just have a nop for anything which likes to process arguments (it's already been processed)

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 7 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:551]: no, if you look at the signature
```haskell
noOneConstantCase ::
  Env.Ops m => Instr.ExpandedOp -> Protect -> m Env.Expanded
```
they only touch the ops, the stack effect happens in the gen function

```haskell
onTwoArgsGen :: OnTerm2Gen m Protect Env.Expanded
onTwoArgsGen applyOperation typ arguments = do
  -- last argument evaled first
  v <- traverse (protect . (inst >=> promoteTopStack)) (reverse arguments)
  case v of
    instr2 : instr1 : _ -> do
      -- Get the result of the operation on the arguments
      res <- applyOperation instr2 instr1
      -- drop the two arguments on the stack
      modify @"stack" (VStack.drop 2)
      -- add the result to the stack
      consVal res typ
      -- return the operation
      pure res
    _ -> throw @"compilationError" Types.NotEnoughArguments

```

namely
```
      modify @"stack" (VStack.drop 2)
```
is what changes the vstack before we say the top is res

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 3 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:471]: yeah I could probably with this, didn't think of it before

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 2 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:61]: Sure I"ll change it to error, I was going to fix it in this pr anyways, since it's just a pair and unit call

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 5 above)
Comment (cwgoes) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:559]: let's either delete it or clarify

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 7 above)
Comment (cwgoes) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:551]: right OK

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 9 above)
Comment (cwgoes) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:834]: it's fine for now

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 10 above)
Comment (cwgoes) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:180]: ah ah I see, yes this is better, thanks!

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs (hunk 2 above)
Comment (mariari) [reply library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs]: both converted, just do a quick look over at the pair and unit translations

**Review by cwgoes — APPROVED** [review]
LGTM.
